modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -1,66 +1,164 @@
-            .map(_ => none[Hashed[GlobalIncrementalSnapshot]])
+            .warn(e)(s"Failure pulling single snapshot with hash=${hash.shortValue}")
+            .as(none)
         }
 
       def pullGlobalSnapshot(ordinal: SnapshotOrdinal): F[Option[Hashed[GlobalIncrementalSnapshot]]] =
         pullGlobalSnapshot(l0GlobalSnapshotClient.get(ordinal)).handleErrorWith { e =>
           logger
-            .warn(e)(s"Failure pulling single snapshot with ordinal=$ordinal")
-            .map(_ => none[Hashed[GlobalIncrementalSnapshot]])
+            .warn(e)(s"Failure pulling single snapshot with ordinal=${ordinal.show}")
+            .as(none)
+        }
+
+      def pullGlobalSnapshots: F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        maybeMajorityPeerIds
+          .fold(pullGlobalSnapshotsFromRandomPeer)(pullGlobalSnapshotsFromMajority)
+          .handleErrorWith { e =>
+            logger
+              .warn(e)(s"Failure pulling global snapshots")
+              .as(
+                List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[LatestSnapshotTuple]
+              )
+          }
+
+      private def pullGlobalSnapshotsFromMajority(
+        majorityPeerIds: NonEmptyList[PeerId]
+      ): F[Either[LatestSnapshotTuple, List[Hashed[GlobalIncrementalSnapshot]]]] =
+        lastGlobalSnapshotStorage.getOrdinal.flatMap {
+          _.fold {
+            pullLatestSnapshotWithMajorityHash(majorityPeerIds).map(_.asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+          } { lastStoredOrdinal =>
+            for {
+              msd <- getLatestMajoritySnapshotData(majorityPeerIds)
+              l0Peers <- globalL0ClusterStorage.getPeers
+              pulled <- pullVerifiedSnapshots(lastStoredOrdinal.next, msd, l0Peers)
+            } yield pulled.asRight[LatestSnapshotTuple]
+          }
+        }
+
+      private def pullLatestSnapshotWithMajorityHash(
+        majorityPeerIds: NonEmptyList[PeerId]
+      ): F[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)] =
+        globalL0ClusterStorage.getPeers
+          .map(nes => Random.shuffle(nes.toSortedSet))
+          .flatMap {
+            _.tailRecM { l0Peers =>
+              l0Peers.headOption
+                .fold(
+                  NoPeersWithMajorityHash.raiseError[F, Either[SortedSet[L0Peer], LatestSnapshotTuple]]
+                ) { l0Peer =>
+                  (for {
+                    msd <- getLatestMajoritySnapshotData(majorityPeerIds)
+                    (snapshot, info) <- pullLatestSnapshotFromPeer(l0Peer)
+                    verified <- verifyLatestSnapshot(snapshot, msd, majorityPeerIds)
+                  } yield Either.cond(verified, (snapshot, info), l0Peers.tail)).handleErrorWith { err =>
+                    logger
+                      .warn(err)(s"Error pulling latest snapshot from peer ${l0Peer.show}")
+                      .as((l0Peers - l0Peer).asLeft[LatestSnapshotTuple])
```

Comment:
can you remove the for comprehension or change the other code not to mix two syntaxes?
```scala
(getLatestMajoritySnapshotData(majorityPeerIds), pullLatestSnapshotFromPeer(l0Peer)).mapN {
  case (msd, (snapshot, info)) => verifyLatestSnapshot(snapshot, msd, majorityPeerIds).map { verified =>
    Either.cond(verified, (snapshot, info), l0Peers.tail)).handleErrorWith { err =>
      logger
        .warn(err)(s"Error pulling latest snapshot from peer ${l0Peer.show}")
        .as((l0Peers - l0Peer).asLeft[LatestSnapshotTuple])
    }
  }
}
```

Reply:
@marcinwadon Why are we favoring consistent syntax over readability? Both are valid and IMO mixing sometimes gives us code that is easier to read.

Reply:
readability is very subjective. something readable for you may be not readable for me and vice versa. hence, I'd keep a single style at least within the same parts of codebase (files at least).